File: app/src/main/java/com/armsnyder/leftorright/XMLRequest.java
package com.armsnyder.leftorright;

import android.util.Log;

import com.android.volley.NetworkResponse;
import com.android.volley.Request;
import com.android.volley.Response;
import com.android.volley.toolbox.HttpHeaderParser;

import org.w3c.dom.Document;
import org.xml.sax.SAXException;

import java.io.ByteArrayInputStream;
import java.io.IOException;

import javax.xml.parsers.DocumentBuilder;
import javax.xml.parsers.DocumentBuilderFactory;
import javax.xml.parsers.ParserConfigurationException;

public class XMLRequest extends Request<Document> {

    private static final String TAG = "XMLRequest";
    private final Object lock = new Object();
    private Response.Listener<Document> listener;

    public XMLRequest(String url, Response.Listener<Document> listener,
                      Response.ErrorListener errorListener) {
        super(Method.GET, url, errorListener);
        this.listener = listener;
    }

    @Override
    protected Response<Document> parseNetworkResponse(NetworkResponse response) {
        Log.d(TAG, "Parsing response");
        DocumentBuilderFactory dbFactory = DocumentBuilderFactory.newInstance();
        try {
            DocumentBuilder dBuilder = dbFactory.newDocumentBuilder();
            Document document = dBuilder.parse(new ByteArrayInputStream(response.data));
            document.normalizeDocument();
            return Response.success(document,
                    HttpHeaderParser.parseCacheHeaders(response));
        } catch (ParserConfigurationException | SAXException | IOException e) {
            throw new RuntimeException(e);
        }
    }

    @Override
    protected void deliverResponse(Document response) {
        Log.d(TAG, "Delivering response");
        Response.Listener<Document> listenerLocal;
        synchronized (lock) {
            listenerLocal = listener;
        }
        if (listenerLocal != null) {
            listenerLocal.onResponse(response);
        }
    }

    @Override
    public void cancel() {
        super.cancel();
        synchronized (lock) {
            listener = null;
        }
    }
}

File: app/src/main/java/com/armsnyder/leftorright/XMLRequest.kt
package com.armsnyder.leftorright

import android.util.Log
import com.android.volley.NetworkResponse
import com.android.volley.Request
import com.android.volley.Response
import com.android.volley.toolbox.HttpHeaderParser
import org.w3c.dom.Document
import java.io.ByteArrayInputStream
import java.util.concurrent.atomic.AtomicBoolean
import javax.xml.parsers.DocumentBuilderFactory

class XMLRequest internal constructor(url: String,
                                      private val listener: Response.Listener<Document>,
                                      errorListener: Response.ErrorListener) :
        Request<Document>(Request.Method.GET, url, errorListener) {
    private val canceled = AtomicBoolean(false)

    override fun parseNetworkResponse(response: NetworkResponse): Response<Document> {
        Log.d(TAG, "Parsing response")
        val dbFactory = DocumentBuilderFactory.newInstance()
        try {
            val document = dbFactory.newDocumentBuilder().parse(ByteArrayInputStream(response.data))
            document.normalizeDocument()
            return Response.success(document, HttpHeaderParser.parseCacheHeaders(response))
        } catch (e: Exception) {
            throw RuntimeException(e)
        }
    }

    override fun deliverResponse(response: Document) {
        Log.d(TAG, "Delivering response")
        if (!canceled.get()) {
            listener.onResponse(response)
        }
    }

    override fun cancel() {
        super.cancel()
        canceled.set(true)
    }

    companion object {
        private const val TAG = "XMLRequest"
    }
}
